### PR TITLE
Support auto doc comment generation

### PR DESCRIPTION
**Pull request by 333fred**
Adds '\n' and '/' to the on-type format list, enabling vscode support for the newly-added omnisharp feature of auto-generating documentation comments. Closes https://github.com/OmniSharp/omnisharp-vscode/issues/8.

### CODE REVIEW

**Comment by 333fred** [comment]
@NTaylorMullen, I think the razor plugin will have to update whatever on-type formatting you have to also call on `/` and `\n`, since the C# extension itself does not get called back on razor files.

@kasecato, this exposes the support I recently added to omnisharp that calls the roslyn functionality for this. I expect that either you or we will get issues about duplicate comments being generated: this generation will only work if the user has on-type formatting turned on. I'm not sure what mechanism you're using, so it's possible that you will still have some functionality this does not (in addition to positioning the cursor in a better location, which this cannot do as it's using the on-type formatting mechanism which cannot control the cursor). But I wanted to give you a heads up before it was merged.

File: test/integrationTests/documentationCommentAutoFormatting.integration.test.ts
Comment (333fred) [inline test/integrationTests/documentationCommentAutoFormatting.integration.test.ts:10]: node.js does not have `replaceAll`, unfortunately...

**Comment by kasecato** [comment]
@333fred This is a feature to output doccoment from AST, right? My extension is completely regular expression, so it was getting harder and harder to maintain... I've been wanting to get the AST from the roslyn for some time. I'll announce the removal of my extension if necessary.

**Comment by 333fred** [comment]
> @333fred This is a feature to output doccoment from AST, right? My extension is completely regular expression, so it was getting harder and harder to maintain... I've been wanting to get the AST from the roslyn for some time. I'll announce the removal of my extension if necessary.

Yep. It's using the same functionality that inserts the doc comments in VS itself.

**Comment by NTaylorMullen** [comment]
> @NTaylorMullen, I think the razor plugin will have to update whatever on-type formatting you have to also call on `/` and `\n`, since the C# extension itself does not get called back on razor files.

We don't currently support OnTypeFormatting in Razor's VSCode implementation so I'd imagine things would continue to work as is. Out of curiosity, how do you all control the cursor when generating doc comments? We've made it work in VS via a custom protocol message; however, that doesn't exist in VSCode AFAIK.

**Review by JoeRobich — APPROVED** [review]

**Comment by 333fred** [comment]
> Out of curiosity, how do you all control the cursor when generating doc comments? We've made it work in VS via a custom protocol message; however, that doesn't exist in VSCode AFAIK.

That's the secret: we do not. It does mean that the cursor will be placed at the end of the generated comment, but I think that's a worthwhile tradeoff to not having the generation at all.

**Comment by NTaylorMullen** [comment]
> That's the secret: we do not. It does mean that the cursor will be placed at the end of the generated comment, but I think that's a worthwhile tradeoff to not having the generation at all.

Gotcha. Ya that's unfortunate in terms of parity but agree that it's better then nothing at all

**Comment by 333fred** [comment]
@JoeRobich do you have any idea what is going on with travis here? I can't actually see any errors?

**Comment by JoeRobich** [comment]
@333fred I hate how the gulp command swallows errors. You can get the error locally by running `tsc -p .`.

```console
~/Source/omnisharp-vscode [auto-doc-gen +0 ~2 -0 !]> tsc -p .
test/integrationTests/documentationCommentAutoFormatting.integration.test.ts:6:1 - error TS6133: 'skip' is declared but its value is never read.

6 import { skip } from 'rxjs/operators';
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error.
```

**Comment by 333fred** [comment]
Alright, now I believe this is just waiting for a new omnisharp-roslyn release.

**Comment by vchirikov** [comment]
It doesn't work on https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.23.8
omnisharp-roslyn: 1.37.6-beta.3
p.s. I have snippets that start with `///`, maybe what's why. Can you try to check with this scenario?
```
"/// <inheritdoc>": {
	"scope": "csharp",
	"prefix": "//i",
	"body": [
		"/// <inheritdoc />",
	],
	"description": "Adds '<inheritdoc />'"
},
```

**Comment by 333fred** [comment]
@vchirikov do you have on-type formatting turned on? This support comes via that mechanism.

**Comment by vchirikov** [comment]
@333fred 
No, I have `"editor.formatOnType": false,`. It's not obliviously that this feature included in `formatOnType`, maybe better to change description in changelog at least.
So, I have to use `formatOnType` or use `k--kato.docomment`, isn't it? :(

**Comment by 333fred** [comment]
> So, I have to use formatOnType or use k--kato.docomment, isn't it?

Correct.

**Pull request merged**
